File: programmers/2020_kakao_internship/경주로 건설.cpp
#include <bits/stdc++.h>
#define INF 0x3f3f3f3f
using namespace std;
int dx[] = {0,0,-1,1};
int dy[] = {-1,1,0,0};
int n;
struct Road { int x, y ,cost, dir; };
int bfs(vector<vector<int>> board){
    int ans = INF;
    queue <Road> q;

    q.push({0,0,0,-1});
    board[0][0] = 1;
    while(!q.empty()){
        int x = q.front().x;
        int y = q.front().y;
        int dir = q.front().dir;
        int cost = q.front().cost;
        q.pop();
        if(x == n-1 && y == n-1) {ans = min(ans,cost); continue;}
        for(int k = 0; k < 4; k++){
            int nx = x + dx[k];
            int ny = y + dy[k];
            int nc = 0;
            if(0 > nx || nx >= n || 0 > ny || ny >= n) continue;
            if(board[nx][ny] == 1) continue;

            if(dir == -1 || dir == k) nc = cost + 100;

            else if(dir != k) nc = cost + 600;

            if(!board[nx][ny] || board[nx][ny] >= nc){
                board[nx][ny] = nc;
                q.push({nx,ny,nc,k});
            }
        }
    }
    for(int i = 0; i < n; i++){
        for(int j = 0; j < n ; j++){
            cout << board[i][j] << ' ';
        }
        cout << '\n';
    }
    return ans;
}

int solution(vector<vector<int>> board) {
    n = board.size();
    return bfs(board);
}

File: programmers/2020_kakao_internship/경주로 건설.js
const dx = [0, 0, -1, 1];
const dy = [-1, 1, 0, 0];
let n;

function bfs(board) {
  let ans = 0x3f3f3f3f;
  let q = [[0, 0, 0, -1]];
  board[0][0] = 1;

  while (q.length !== 0) {
    const x = q[0][0];
    const y = q[0][1];
    const cost = q[0][2];
    const dir = q[0][3];
    q.shift();
    // if (x == n - 1 && y == n - 1) {
    //   ans = Math.min(ans, cost);
    //   continue;
    // }
    for (let k = 0; k < 4; k++) {
      const nx = x + dx[k];
      const ny = y + dy[k];
      let nc = 0;
      if (0 > nx || nx >= n || 0 > ny || ny >= n) continue;
      if (board[nx][ny] === 1) continue;

      if (dir === -1 || dir === k) nc = cost + 100;
      else if (dir !== k) nc = cost + 600;

      if (board[nx][ny] === 0 || board[nx][ny] >= nc) {
        board[nx][ny] = nc;
        q.push([nx, ny, nc, k]);
      }
    }
  }
  return board[n - 1][n - 1];
}

function solution(board) {
  n = board.length;
  return bfs(board);
}
